maze_transformer/training/config.py

```diff
@@ -210,7 +216,81 @@ def transformer_config(self) -> HookedTransformerConfig:
             d_vocab=len(self.dataset_cfg.token_arr),
         )
 
+    def transformer_config(self) -> HookedTransformerConfig:
+        warnings.warn(
+            "transformer_config is deprecated, use hooked_transformer_cfg instead",
+            DeprecationWarning,
+        )
+        return self.hooked_transformer_cfg
+
     def create_model(self) -> HookedTransformer:
         return HookedTransformer(
-            cfg=self.transformer_config(), tokenizer=self.tokenizer
+            cfg=self.hooked_transformer_cfg,
+            tokenizer=self.tokenizer,
+        )
+
+    def create_model_zanj(self) -> ZanjHookedTransformer:
+        return ZanjHookedTransformer(self)
+
+
+@set_config_class(ConfigHolder)
+class ZanjHookedTransformer(ConfiguredModel, HookedTransformer):
+    """A hooked transformer that is configured by a ConfigHolder
+
+    the inheritance order is critical here -- super() does not call parent, but calls the next class in the MRO
+    So, we need ConfiguredModel to take the ConfigHolder and pass kwargs to HookedTransformer
+    """
+
+    def __init__(self, cfg_holder: ConfigHolder) -> None:
+        super().__init__(
+            # for ConfiguredModel
+            zanj_model_config=cfg_holder,
+            # for HookedTransformer
+            cfg=cfg_holder.hooked_transformer_cfg,
+            tokenizer=cfg_holder.tokenizer,
+        )
+
+    def _load_state_dict_wrapper(
+        self,
+        state_dict: dict[str, Any],
+        **kwargs,
+    ) -> None:
+        """this is a wrapper around the _load_state_dict function that allows us to do extra things when loading a state dict"""
+
+        recover_exact: bool = kwargs.get("recover_exact", False)
+        fold_ln: bool = kwargs.get("fold_ln", True)
+
+        if self.zanj_model_config.model_cfg.are_layernorms_folded and fold_ln:
+            raise ValueError(
+                "Cannot fold layernorms twice! the saved model already has layernorms folded"
+            )
+
+        if recover_exact and fold_ln:
+            raise ValueError(
+                "Can't recover exact weights if the layernorm is to be folded!"
+            )
+
+        self.zanj_model_config.model_cfg.are_layernorms_folded = fold_ln
+        self.zanj_model_config.model_cfg.are_weights_processed = not recover_exact
+
+        self.load_and_process_state_dict(
+            state_dict,
+            fold_ln=False,
+            center_writing_weights=not recover_exact,
+            center_unembed=not recover_exact,
+            refactor_factored_attn_matrices=not recover_exact,
```

Comment:
Are we sure we always want to set these 3 props to the same value? In the hackathon, `refactor_factored_attn_matrices=True` caused my model to perform terribly, but everything was fine if I turned it off and left the other 2 on. The TransformerLens docs note that it's an experimental method.

Reply:
Was not aware of this issue, I inherited the code from the old `load_model_with_config` which had all those hardcoded to true. @afspies, what would make the most sense here for the research team?

Reply:
I was under the impression that none of these affect the computation of the model. It seems this is only true for `center_writing_weights` and partially for `center_unembed`. 

`refactor_factored_attn_matrices` seems very involved and it is not clear to me that it is something we should be using.

I would probably have separate flag for `refactor_factored_attn_matrices` and the other two. It might also be reasonable to just have `center_writing_weights` and `center_unembed` true based on an `eval_flag` as it's not the case that they aren't `exact`

Reply:
so tentatively I have as defaults:
- `recover_exact = False`
- `fold_ln = False`
- `refactor_factored_attn_matrices = False`

and `recover_exact` disables `center_writing_weights` and `center_unembed` if set to true

Alex, what do you mean by "it's not the case that they aren't exact"?

Reply:
It's probably fine to have the defaults as False, although layernorm folding makes a lot of sense as a default practice.

It should not be the case that `center_writing_weights` or `center_unembed` will change the outputs of the model for a given input (though `center_unembed` will change logits before softmax) and so you aren't really changing the computation performed by the model. Perhaps by "exact" we mean "exactly the same way of expressing the parameters as when the model was saved", so it's not really a problem.

Reply:
yeah, `recover_exact` in this context means "make it such that the state dicts are exactly equal". we have two different comparison method in the tests, one which explicitly compares the config and state dict plus another which just compares config + output equality.